Consumption: introduce new cmdlets to CRUD budget, Get price sheet, marketplace, reservation summary/detail

bgsky opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

Introduce 4 new cmdlets to CRUD budget. Budget can be managed by the user to help monitor their cost or usage.
Introduce 4 new cmdlets to get price sheet, marketplace, reservation summary and reservation detail

## Checklist

- [X ] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [X ] The title of the PR is clear and informative
- [X ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [X ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [X ] If applicable, the changes made in the PR have proper test coverage
- [X ] For public API changes to cmdlets:
    - [X ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [X ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

bgsky commented:
Please refer to an issue https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/69

bgsky commented:
@maddieclayton @cormacpayne @ms-premp Follow up with new cmdlets on budget, this pull request also introduces another 4 new cmdlets on price sheet, marketplace, and reservation summary/detail. Please refer to review request on [budget](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/69), [price sheet & marketplace](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/70) and [reservation summary/detail](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/76), and let us know about any comments.

maddieclayton commented:
@bgsky Looks like the build is failing for help file issues, please check out here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/4927/artifact/src/Package/ValidateHelpIssues.csv

bgsky commented:
@maddieclayton @ms-premp @shafqatanwar Thanks for your feedback on missing headers on help files. We update them by adding online links, please review and lets know if latest static analysis failure is on our side.

maddieclayton commented:
@bgsky Please check out the "SignatureIssues.csv" file here: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/4935/, and fix issues.  Please also take a look at the BreakingChanges.csv file - if you can confirm that each of these changes cannot cause a script to break (values outside of the range would previously fail on the server side and now are caught on the powershell side), you can suppress the issues following this doc: https://github.com/Azure/azure-powershell/blob/preview/documentation/Debugging-StaticAnalysis-Errors.md.  If however, these changes could break a previously working script, we cannot allow them in a non-breaking change release.  The next breaking change release will be in November.

bgsky commented:
@maddieclayton Following your suggestions, we have fixed the build errors on breaking changes by removing those range validations. Plan to add them back before November as part of breaking change release. 

Please review and proceed on your side, and lets know if there is any concern. We really need it public in next milestone around mid June.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Reservation/GetAzureRmConsumptionReservationSummary.cs:
```diff
@@ -0,0 +1,92 @@
+                        ConsumptionManagementClient.ReservationsSummaries.ListByReservationOrder(ReservationOrderId,
+                            Grain, filter);
+                }
+            }
```
paging?

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/AzureRM.Consumption.md:
```diff
@@ -14,3 +14,35 @@ The topics in this section document the Azure PowerShell cmdlets for Azure Consu
 ### [Get-AzureRmConsumptionUsageDetail](Get-AzureRmConsumptionUsageDetail.md)
 Get usage details of the subscription.
 
+## AzureRM.Consumption Cmdlets
```
Remove this line.

bgsky replied:
Remove all extra lines of `AzureRM.Consumption Cmdlets`

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Get-AzureRmConsumptionBudget.md:
````diff
@@ -0,0 +1,116 @@
+
+### Example 1
+```powershell
+PS C:\> Get-AzureRmConsumptionBudget
````
Please provide sample output for all examples.  Apply across all help files.

bgsky replied:
Add sample output across all help files

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Get-AzureRmConsumptionBudget.md:
```diff
@@ -0,0 +1,116 @@
+
+## OUTPUTS
+
+### System.Collections.Generic.List`1[[Microsoft.Azure.Commands.Consumption.Models.PSBudget, Microsoft.Azure.Commands.Consumption, Version=0.3.3.0, Culture=neutral, PublicKeyToken=null]]
```
Change this to just "Microsoft.Azure.Commands.Consumption.Models.PSBudget".  Apply to all help files.

bgsky replied:
Fixed it by fixing cmdlet file

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Remove-AzureRmConsumptionBudget.md:
````diff
@@ -0,0 +1,148 @@
+```
+
+### -PassThru
+{{Fill PassThru Description}}
````
Fill in this description

bgsky replied:
Fixed it by fixing cmdlet file

bgsky commented:
@maddieclayton Thanks for your feedback. We are working on changes suggested. For example, will generate and destroy budget for each test. 

One question on resource group name and budget name generated in test time to re-record tests. Wonder who is going to re-record. Some test cases with filters (on resource group, resource, or meter) apply only on our test subscription and/or resource group scope.

maddieclayton commented:
@bgsky Sometimes on our backend changes are made that require tests to be rerecorded, and we prefer to do it if possible.  Additionally, it will make any changes you make in the future easier as you will not have to figure out how to properly set up a subscription before testing.  If there is setup that cannot be done in PowerShell, please leave detailed instructions on how to properly set up the subscription before recording the test.

bgsky commented on src/ResourceManager/Consumption/AzureRM.Consumption.Netcore.psd1:
```diff
@@ -12,7 +12,7 @@
 
 # Version number of this module.
-ModuleVersion = '0.11.0'
+ModuleVersion = '0.12.0'
```
reverted

bgsky commented on src/ResourceManager/Consumption/AzureRM.Consumption.psd1:
```diff
@@ -12,7 +12,7 @@
 
 # Version number of this module.
-ModuleVersion = '0.3.3'
+ModuleVersion = '0.4.0'
```
reverted

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+#>
+function Test-NewBudget
+{
+    $budget = New-AzureRmConsumptionBudget -Amount 60 -Name PSBudget -Category Cost -StartDate 2018-06-01 -EndDate 2018-11-01 -TimeGrain Monthly
```
Budget tests are reorganized to create and destroy budget.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+#>
+function Test-NewBudgetAtResourceGroupLevel
+{
+    $budget = New-AzureRmConsumptionBudget -ResourceGroupName RGBudgets -Amount 60 -Name PSBudgetRG -Category Cost -StartDate 2018-06-01 -EndDate 2018-11-01 -TimeGrain Monthly 
```
We would like to hard code resource group name `RGBudgets` instead of depending on another cmdlet Get-AzureRmResourceGroup to verify budget test. Even created under test subscription, API works if resource group name is not under the subscription. Rerecording can proceed, although current spend in this situation will be null.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+    $budget = New-AzureRmConsumptionBudget -Amount 60 -Name PSBudget -Category Cost -StartDate 2018-06-01 -EndDate 2018-11-01 -TimeGrain Monthly
+	Assert-NotNull $budget
+	Assert-AreEqual 60 $budget.Amount
+	Assert-AreEqual PSBudget $budget.Name
```
Apply `getAssetName` to get budget name and notification key

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+#>
+function Test-GetBudgets
+{
+    $budgets = Get-AzureRmConsumptionBudget 
```
Budget tests are reorganized to create and destroy budget.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+#>
+function Test-GetBudgetsAtResourceGroupLevel
+{
+    $budgets = Get-AzureRmConsumptionBudget -ResourceGroupName RGBudgets
```
Budget tests are reorganized to create and destroy budget.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+#>
+function Test-GetBudgetByName
+{
+    $budget = Get-AzureRmConsumptionBudget -Name PSBudget
```
Budget tests are reorganized to create and destroy budget.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+    $budget = Get-AzureRmConsumptionBudget -Name PSBudget
+	Assert-NotNull $budget
+	Assert-AreEqual PSBudget $budget.Name
+	Assert-NotNull $budget.Amount
```
Validate amount by specific value

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,237 @@
+	$response = Remove-AzureRmConsumptionBudget -Name PSBudget -PassThru
+	Assert-AreEqual True $response
+}
+
```
Add Write-Debug on each test cmdlet call of budget

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/MarketplaceTests.ps1:
```diff
@@ -0,0 +1,181 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ----------------------------------------------------------------------------------
+
```
Cmdlet can return data if the subscription has usage data in Azure Marketplace. If not, it return empty data.

maddieclayton replied:
Okay, please add instructions to the top of any test that cannot be rerecorded automatically on any subscription.  This could be instructions on things to do in a random subscription to get the correct setup, or the subscription that this needs to be recorded on (and who to contact to either get access to the subscription or to ask to rerecord).  This will make it easier for us to deal with the possible need to rerecord any of the tests.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/PriceSheetTests.ps1:
```diff
@@ -0,0 +1,100 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ----------------------------------------------------------------------------------
+
```
Most subscriptions will have price sheet data, if some resources are created under it and each resource has its pricing represented by a row in price sheet.

maddieclayton replied:
Okay please add instructions to these test too then.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/ReservationTests.ps1:
```diff
@@ -0,0 +1,167 @@
+List monthly reservation summaries with reservation order Id
+#>
+function Test-ListReservationSummariesMonthlyWithOrderId
+{
```
It works for a subscription if there is a reversed instance created under it.  Follow their [cmdlet link](https://github.com/Azure/azure-powershell/tree/1abd90669ede17bac9288768a3e69bfdf7ef855d/src/ResourceManager/Reservations/Commands.Reservations/Cmdlets), we don't have direct cmdlet access to id or order id unless from portal. It seems they works like subscription guid.

maddieclayton replied:
Got it.  Then all you need to do is add instructions about how to record the test here as well.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/GetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,88 @@
+            }
+            catch (ErrorResponseException e)
+            {
+                WriteWarning(e.Body.Error.Message);
```
Fix by throwing exception

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/GetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,88 @@
+
+            if (budget != null)
+            {
+                WriteObject(new List<PSBudget> {new PSBudget(budget)}, true);
```
Fix as single PSBudget object

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/NewAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,225 @@
+        [ValidateNotNullOrEmpty]
+        public string[] MeterFilter;
+
+        [Parameter(Mandatory = true, HelpMessage = "Name of a budget.")]
```
Shuffle all mandatory parameters on top

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/NewAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,225 @@
+            }
+            catch (ErrorResponseException e)
+            {
+                WriteWarning(e.Body.Error.Message);
```
Fix by throwing exception

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/RemoveAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,66 @@
+    [Cmdlet(VerbsCommon.Remove, "AzureRmConsumptionBudget", SupportsShouldProcess = true)]
+    [OutputType(typeof(bool))]
+    public class RemoveAzureRmConsumptionBudget : AzureConsumptionCmdletBase
+    {
```
@maddieclayton We don't have piping implemented in budget, as we have budget name as the key instead of budget object. Remove or Set cmdlet asks for same parameter as Get cmdlet, and piping in Remove seems redundant in this scenario. Lets know if piping is required for all CRUD cmdlets

maddieclayton replied:
@bgsky Does the budget object have the budget name as a property? If this is the case, when you implement the InputObject parameter set, you will just extract the budget name from the object.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/RemoveAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,66 @@
+            }
+            catch (ErrorResponseException e)
+            {
+                WriteWarning(e.Body.Error.Message);
```
Fix by throwing exception

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/SetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,262 @@
+
+                if (notification.ContactEmails != null)
+                {
+                    contactCount += notification.ContactEmails.Count;
```
A bug here, fix logic

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/SetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,262 @@
+
+                if (notifications.Keys.Count >= 6)
+                {
+                    WriteWarning("Budget can only have up to five notifications.");
```
Fix by throwing an exception

maddieclayton replied:
Still see the warning.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/GetAzureRmConsumptionMarketplace.cs:
```diff
@@ -0,0 +1,204 @@
+
+namespace Microsoft.Azure.Commands.Consumption.Cmdlets
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmConsumptionMarketplace"), OutputType(typeof(List<PSMarketplace>))]
```
Fix output type by single PSMarketplace object

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/GetAzureRmConsumptionPriceSheet.cs:
```diff
@@ -0,0 +1,121 @@
+
+        [Parameter(Mandatory = false, HelpMessage = "Expand the price sheets based on MeterDetails.")]
+        [ValidateNotNullOrEmpty]
+        public string Expand { get; set; }
```
Add ValidateSet

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/GetAzureRmConsumptionPriceSheet.cs:
```diff
@@ -0,0 +1,121 @@
+            }
+
+            int numberToFetch = MaxNumberToFetch;
+            if (this.Top.HasValue && this.Top.Value < numberToFetch)
```
Remove `this.Top.Value < numberToFetch`

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/GetAzureRmConsumptionMarketplace.cs:
```diff
@@ -0,0 +1,204 @@
+    [Cmdlet(VerbsCommon.Get, "AzureRmConsumptionMarketplace"), OutputType(typeof(List<PSMarketplace>))]
+    public class GetAzureRmConsumptionMarketplace : AzureConsumptionCmdletBase
+    {
+        private const int MaxNumberToFetch = 1000;
```
Used near the end of parameter block in ExecuteCmdlet method

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Reservation/GetAzureRmConsumptionReservationDetail.cs:
```diff
@@ -0,0 +1,79 @@
+    {
+        [Parameter(Mandatory = true, HelpMessage = "The end data (YYYY-MM-DD in UTC) of the reservation detail.")]
+        [ValidateNotNullOrEmpty]
+        public DateTime EndDate { get; set; }
```
This is mandatory, so skip `?` check

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Reservation/GetAzureRmConsumptionReservationDetail.cs:
```diff
@@ -0,0 +1,79 @@
+                {
+                    reservationDetails =
+                        ConsumptionManagementClient.ReservationsDetails.ListByReservationOrder(ReservationOrderId, filter);
+                }
```
We don't support paging in API implementation in reservation summary and detail, so skip here

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Reservation/GetAzureRmConsumptionReservationSummary.cs:
```diff
@@ -0,0 +1,92 @@
+            string filter = null;
+            if (this.Grain.Equals("daily", StringComparison.CurrentCultureIgnoreCase))
+            {
+                var from = this.StartDate?.ToString(Constants.Formats.DateTimeParameterFormat);
```
API will throw an exception if not provided

maddieclayton replied:
Can you throw a nice exception here? Guessing the error from the API is not as nice as one you could write.

maddieclayton replied:
Sure, please just be sure that the error thrown by the API will make sense to the user.

bgsky commented:
@maddieclayton @ms-premp @shafqatanwar We resolved and replied to individual comments, also summarized below. Please review the latest iteration and lets know if you have any comments.
1, reverted module version back
2, added notification parameter set on budget 
3, added a few parameter validation
4, changed output type from list to single PS object
5, changed from WriteWarning to WriteException in budget cmdlet
6, shuffled mandatory parameter to top
7, added sample output across help files
8, created and destroyed budget object to be indepedent at test time

Answers to questions from you:
1, Reservation cmdlets work for a subscription if a reserved instance is created under it. Id and order id can be found in portal.
2, Marketplace cmdlet works for a subscription if there is a resource created under it through Azure Marketplace.
3, PriceSheet cmdlet works for most subscriptions if there is a resource with meter created under it so Azure pricing will apply to those subscriptions.

Questions from us:
1, We hard coded resource group name instead of calling Get-AzureRmResourceGroup. API supports rerecording if resource group is not under subscription, while current spend will be null in this case. Lets know if it works on your side of rerecording.
2, We don't have piping implemented in budget, as we have budget name as the key instead of budget object. Remove or Set cmdlet asks for same parameter as Get cmdlet, and piping in Remove seems redundant in this scenario. Lets know if piping is required for all CRUD cmdlets.

maddieclayton commented on src/ResourceManager/Consumption/AzureRM.Consumption.Netcore.psd1:
```diff
@@ -12,7 +12,7 @@
 
 # Version number of this module.
-ModuleVersion = '0.12.0'
+ModuleVersion = '0.11.0'
```
I think it was bumped for a new release version.  We take care of versioning in all cases, so please revert this.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,143 @@
+	$budgetSet3 = Set-AzureRmConsumptionBudget -Name $budgetName -NotificationKey $notificationKey -NotificationDisable
+	Assert-NotNull $budgetSet3
+	Assert-AreEqual $budgetName $budgetSet3.Name
+	Assert-AreEqual 1 $budgetSet3.Notification.Count
```
Can you add more assert to all of these sets to ensure the proper properties are set?

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,143 @@
+	$budgetName = Get-BudgetName
+	$notificationKey = Get-NotificationKey
+	# This resource group is created at the scope of test subscription
+	$resourceGroupName = "CriTest"
```
Please use "getAssetName" here as well, and create the resourcegroup at the beginning and destroy it at the end.  Unless there is special information you need the resource group to have, it seems like there is no reason to hardcode the resource group here.

maddieclayton replied:
Just saw your comment - if the resourcegroup does not need to exist here, then you can just change this to getAssetName and not create/destroy it.

bgsky replied:
@maddieclayton Will call New-AzureRmResourceGroup to create and later destroy it in testing

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,143 @@
+	$notificationKey1 = Get-NotificationKey
+	$notificationKey2 = Get-NotificationKey
+	# This resource group is created at the scope of test subscription
+	$resourceGroupName = "RGBudgets"
```
Same comment here about creating and destroying the resource group, and using getAssetName

maddieclayton replied:
I think this one is actually necessary because you pass in -ResourceGroupName in the "New" cmdlet. Let me know if I am wrong and the resourcegroup does not need to actually exist (which seems weird).

bgsky replied:
@maddieclayton Will call New-AzureRmResourceGroup to create and later destroy it in testing

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,143 @@
+#>
+function Test-GetBudgets
+{
+    $budgets = Get-AzureRmConsumptionBudget 
```
Please create at least one budget before running this so that is will always pass (and then destroy it after the call.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption.Test/ScenarioTests/BudgetTests.ps1:
```diff
@@ -0,0 +1,143 @@
+	Assert-AreEqual 2 $budgetSet2.Notification.Count
+
+	Write-Debug "Remove the budget $budgetName"
+	$response = Remove-AzureRmConsumptionBudget -Name $budgetName -ResourceGroupName $resourceGroupName -PassThru
```
Can you add a "Get-AzureRmConsumptionBudget -Name $budgetName" and asset that is returns nothing? That is the best way to test a remove.

bgsky replied:
@maddieclayton We couldn't add another `Get` cmdlet here, because API will throw an exception insteading of returning nothing.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/NewAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,243 @@
+        [ValidateNotNullOrEmpty]
+        public string[] ResourceGroupFilter;
+
+        [Parameter(ParameterSetName = ParameterSetNames.SubscriptionItemParameterSet, Mandatory = false, HelpMessage = HelpMessages.NotificationKey)]
```
I realized that I was probably not explicit enough about this.  To do the parameter sets properly, please remove all of the notification related parameters (basically any that require notification key to work correctly) from the SubscriptionItemParameterSet.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/SetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,276 @@
+
+namespace Microsoft.Azure.Commands.Consumption.Cmdlets.Budget
+{   
+    using Budget = Management.Consumption.Models.Budget;
```
Same comment here about InputObject and ResourceId parameter sets.

bgsky replied:
We don't apply piping to SetBudget cmdlet, because the object from GetBudget is a model of PSBudget, and it cannot be converted to budget model as some parameters in budget model are private set. We would like to skip piping in SetBudget cmdlet.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/SetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,276 @@
+
+        [Parameter(ParameterSetName = ParameterSetNames.SubscriptionItemParameterSet, Mandatory = false, HelpMessage = HelpMessages.NotificationDisabled)]
+        [Parameter(ParameterSetName = ParameterSetNames.NotificationItemParameterSet, Mandatory = false, HelpMessage = HelpMessages.NotificationDisabled)]
+        public SwitchParameter NotificationDisabled;
```
What happens if the user gives both Enabled and Disabled? This is not a pattern PowerShell cmdlets follow, please follow the original comment: Only have one of either "Enable" or "Disable", and if the user doesn't provide the parameter, reset to default.  If this cmdlet is really a "Patch", let me know as the advice is different.

bgsky replied:
Disable was removed.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/SetAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,276 @@
+        [ValidateNotNullOrEmpty]
+        public string[] ResourceGroupFilter;
+
+        [Parameter(ParameterSetName = ParameterSetNames.SubscriptionItemParameterSet, Mandatory = false, HelpMessage = HelpMessages.NotificationKey)]
```
Fix the notification/subscription parameter sets as I described in "New"

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/GetAzureRmConsumptionPriceSheet.cs:
```diff
@@ -0,0 +1,120 @@
+
+        [Parameter(Mandatory = false, HelpMessage = "Expand the price sheets based on MeterDetails.")]
+        [ValidateNotNullOrEmpty]
+        [ValidateSet("MeterDetails")]
```
Is this going to have more values eventually? Because otherwise it seems like this parameter should just be a switch parameter "ExpandMeterDetails"

bgsky replied:
We don't have plans to add more values yet. So make it `ExpandMeterDetail`

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/GetAzureRmConsumptionPriceSheet.cs:
```diff
@@ -0,0 +1,120 @@
+            }
+            catch (ErrorResponseException e)
+            {
+                WriteWarning(e.Body.Error.Message);
```
Change to error (apply to all cmdlets)

bgsky replied:
Changed

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Get-AzureRmConsumptionBudget.md:
````diff
@@ -0,0 +1,144 @@
+
+### Example 1: Get a list of budgets at subscription level
+```powershell
+PS C:\> Get-AzureRmConsumptionBudget
````
Can you fix the formating for the examples? If you go look at the md file "view" option, you'll see that the formatting is off.

bgsky replied:
Formatting was fixed across help files.

maddieclayton commented on src/ResourceManager/Consumption/Commands.Consumption/help/Get-AzureRmConsumptionUsageDetail.md:
```diff
@@ -268,7 +330,7 @@ For more information, see about_CommonParameters (http://go.microsoft.com/fwlink
 ## OUTPUTS
 
-### System.Collections.Generic.List`1[[Microsoft.Azure.Commands.Consumption.Models.PSUsageDetail, Microsoft.Azure.Commands.Consumption, Version=0.3.2.0, Culture=neutral, PublicKeyToken=null]]
+### System.Collections.Generic.List`1[[Microsoft.Azure.Commands.Consumption.Models.PSUsageDetail, Microsoft.Azure.Commands.Consumption, Version=0.3.3.0, Culture=neutral, PublicKeyToken=null]]
```
Fix to be just "Microsoft.Azure.Commands.Consumption.Models.PSUsageDetail"

bgsky replied:
Change to single PSUsageDetail results in build failure. Lets know if this is a breaking change.

bgsky commented:
@maddieclayton We resolved most comments. Please review and lets know if build failure is related to consumption module. We would like to publish the module in 2018-06-15 milestone. Also lets know if there is any concern on that.

A few budget issues are not yet resolved. Get-AzureRmResourceGroup cmdlet got an error, so it is still hard coded. We don't apply piping to SetBudget cmdlet, because the object from GetBudget is a model of PSBudget, and it cannot be converted to budget model as some parameters in budget model are private set. We would like to skip piping in SetBudget cmdlet. 

We plan to fix those issues after publishing in 2018-06-15 milestone, as the users are requesting those cmdlets. Please lets know if any concern on your side.

bgsky commented:
@maddieclayton According to staticanalysis file, build error results from an output change of Get-AzureRmConsumptionUsageDetail from List<PSUsageDetail> to PSUsageDetail. Please lets know if this change is allowed.

bgsky commented on src/ResourceManager/Consumption/Commands.Consumption/Cmdlets/Budget/NewAzureRmConsumptionBudget.cs:
```diff
@@ -0,0 +1,240 @@
+                notifications: CreateNotifications());
+
+                Budget responseBudget = null;
+                try
```
@maddieclayton Already have a wrapper on top with ShouldProcess. Is it necessary?

maddieclayton commented:
https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/928/
